PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+    distances_shared = np.ndarray(distances.shape, dtype = distances.dtype, buffer = shm_distances.buf)
+    distances_shared[:] = distances[:]
+
+    shm_distance_ranks = shared_memory.SharedMemory(create = True, size = distance_ranks.nbytes, name = 'shm_distance_ranks')
+    distance_ranks_shared = np.ndarray(distance_ranks.shape, dtype = distance_ranks.dtype, buffer = shm_distance_ranks.buf)
+    distance_ranks_shared[:] = distance_ranks[:]
+
+    isolate_list_shared = shared_memory.ShareableList(isolate_list, name = 'shm_isolate_list')
+    
+    # run clustering for an individual R
+    if num_processes == 1:
+        for R in rank_list:
+            lineage_clustering[R], lineage_seed[R], neighbours[R], previous_lineage_clustering[R] = run_clustering_for_R(R,
+                                                                                        null_cluster_value = null_cluster_value,
+                                                                                        qlist = qlist,
+                                                                                        existing_scheme = existing_scheme,
+                                                                                        distances_length = distances.shape,
+                                                                                        distances_type = distances.dtype)
+
+
+    else:
+
+        with Pool(processes = num_processes) as pool:
+            results = pool.map(partial(run_clustering_for_R,
+                                null_cluster_value = null_cluster_value,
+                                qlist = qlist,
+                                existing_scheme = existing_scheme,
+                                distances_length = distances.shape,
+                                distances_type = distances.dtype),
+                                rank_list)
+        
+        for n,result in enumerate(results):
+            R = rank_list[n]
+            lineage_clustering[R], lineage_seed[R], neighbours[R], previous_lineage_clustering[R] = result
+        
+    # manage memory
+    shm_distances.close()
+    shm_distances.unlink()
+    del shm_distances
+    shm_distance_ranks.close()
+    shm_distance_ranks.unlink()
+    del shm_distance_ranks
+    isolate_list_shared.shm.close()
+    isolate_list_shared.shm.unlink()
+    del isolate_list_shared
+
+    # store output
+    with open(output + "/" + output + '_lineageClusters.pkl', 'wb') as pickle_file:
+        pickle.dump([lineage_clustering, lineage_seed, neighbours, rank_list], pickle_file)
+    
+    # print output
+    combined = {}
+    titles_list = ['Lineage_R' + str(R)  for R in rank_list]
+    lineage_output_name = output + "/" + output + "_lineage_clusters.csv"
+    with open(lineage_output_name, 'w') as lFile:
+        # print header
+        lFile.write('Id')
+        for t in titles_list:
+            lFile.write(',' + t + '__autocolor')
```

Comment:
Assumes microreact, but tbh I'm fine with that

Reply:
Fixed in 2d712d2.